File: src/java/org/javaturk/dp/ch05/adapter/electricity/powerAdapter2/USTurkishPowerAdapter.java
package org.javaturk.dp.ch05.adapter.electricity.powerAdapter2;

import org.javaturk.dp.ch05.adapter.electricity.domain.tr.TurkishPowerSource;
import org.javaturk.dp.ch05.adapter.electricity.domain.us.USPowerSource;

/**
 * This adapter provides some extra functionality.
 * @author akin
 *
 */
public class USTurkishPowerAdapter implements TurkishPowerSource {
	private USPowerSource usPowerSource;
	
	public USTurkishPowerAdapter(USPowerSource usPowerSource){
		System.out.println("\nUSTurkishPowerAdapter: Converting from USPowerSource to TurkishPowerSource");
		this.usPowerSource = usPowerSource;
		convert110To220(usPowerSource);
	}

	@Override
	public void providePowerAt220V() {
		usPowerSource.providePowerAt110V();
		convert110To220(usPowerSource);
	}

	@Override
	public void turnOn() {
		usPowerSource.pushSwitch();
	}

	@Override
	public void turnOff() {
		usPowerSource.pushSwitch();
	}
	
	private void convert110To220(USPowerSource usPowerSource){
		check();
		System.out.println("USTurkishPowerAdapter: Converting from 110V to 220V");
		regulateVoltage();
	}
	
	/**
	 * Some extra functionality the adaptor provides.
	 */
	private void check(){
		System.out.println("Making some safety checks.");
	}
	
	/**
	 * Some extra functionality the adaptor provides.
	 */
	private void regulateVoltage(){
		System.out.println("Regulating the voltage.");
	}
}

File: src/typescript/com/gokselkucuksahin/dp/structural-patterns/adapter/electricity/powerAdapter2/us-turkish-power-adapter.ts
import {TurkishPowerSource} from "../domain/tr/turkish-power-source";
import {USPowerSource} from "../domain/us/us-power-source";

export class USTurkishPowerAdapter implements TurkishPowerSource {
  private usPowerSource: USPowerSource;
  private on: boolean;

  constructor(usPowerSource: USPowerSource) {
    console.log("\nUSTurkishPowerAdapter: Converting from USPowerSource to TurkishPowerSource");
    this.usPowerSource = usPowerSource;
    this.convert110To220(usPowerSource);
  }

  providePowerAt220V(): void {
    this.usPowerSource.providePowerAt110V();
    this.convert110To220(this.usPowerSource);
  }

  turnOff(): void {
    if (this.on) {
      this.usPowerSource.pushSwitch();
      this.on = false;
    }
  }

  turnOn(): void {
    if (!this.on) {
      this.usPowerSource.pushSwitch();
      this.on = true;
    }
  }

  private convert110To220(usPowerSource: USPowerSource): void {
    this.check();
    console.log("USTurkishPowerAdapter: Converting from 110V to 220V");
    this.regulateVoltage();
  }

  private check(): void {
    console.log("Making some safety checks.");
  }

  private regulateVoltage(): void {
    console.log("Regulating the voltage.");
  }

}